### DISCUSSION TITLE
Better enum support

### DISCUSSION THREAD

**Discussion by fischerscode** [serverpod/serverpod#309] [Ideas]
I think it would be cool if tables could have fields of an type specified as an enum. (At least that's what I tried first...)

This would also be nice to use them directly in expressions.

While it is actually not 100% necessary it might improve the experience.

**What do you think?**

## Example
```
class: LogEntry
table: serverpod_log
fields:
...
  logLevel: LogLevel # currently you have to use int
...
```
```
enum: LogLevel
values:
  - debug
  - info
  - warning
  - error
  - fatal
```
```
var logs = LogEntry.find(session,where: (t) => t.logLevel.equals(LogLevel.error));
```
---
If wanted, I might implement it.

**Comment by vlidholt** [comment]
If you want to look into it, it would need to happen before the 1.0 release, because it is a breaking change. I think it may not be a super easy fix though as quite a few layers are involved (CLI tooling, database, serialization, etc).

**Reply by fischerscode** [reply]
So in case I would provide a PR that changes the "Json - Enum" behavior to one based on a single int value before the 1.0 release, you are fine with it and it might be merged?
In that I'll create an issue, and try to find some time to implement it soon.

About the time window I have... You wrote 1.0 might happen soon. A "next week soon" or would a PR at the end of next week, possibly the weekend be fine?

**Reply by vlidholt** [reply]
Thank you for the contribution! :) I merged it now, added some extra tests too, just to be sure.

**Reply by fischerscode** [reply]
About using enums as described above...

My current goals:
1. Store enums as integers in the database
2. Serialize enums as integers in the api (I think this would esp. be nice if one uses serverpod as an generic http api.)

I tinkered around with some ideas how to achieve the goals an so far I always ended up doing major changes to the serialization system. Since the current system really relies on getting the classname in the json.

Currently I wonder weather it would be better to keep the cross device serialization in the generated `Endpoints` class, and the database serialization in the `TableRow` extending classes, to support integer based enums for everything that faces the outside of serverpod. So `SerializationManager` would only be used for internal storing of objects.

However... I'll probably change and refactor a hole lot of it while keeping new types (something like nested Lists) in mind. In case I succeed, I'll just create a PR and then it'll be your choice whether you want to merge the changes or not. And to be honest, I'd totally understand if you'd decide to keep your logic, but for me it's now a coding challenge.

**Reply by vlidholt** [reply]
I would probably have done some things different if I had rewritten the serialization from scratch. 😅 There are some shortcomings like the ones you are mentioning, like the lack of nested List and Map. One of my initial goals was to have the APIs being more standard REST/JSON, by the way ints are passed, etc. In hindsight, it may have been better to scrap that requirement as the serialized objects are quite complex anyway and 99% of users will use the generated client files anyway. Also, serialization is currently taking place in different places in the code, it would have been nicer to keep it more centralized.

My plan was to perhaps build a new, cleaner, serialization layer for the 2.0 release. Possibly compatible with Protobuf or some other binary protocol.

**Reply by fischerscode** [reply]
So would you prefer a json based serialization or something binary like protobuf?

In case you prefere json: Would you prefer wrapping classes like it is currently done or would you prefer it being done without wrapping? So far, I think both would be possible, (except for thrown objects in endpoints, are they actually currently supported?), but doing it without wrapping should be more efficient.

My current approach would keep the context in mind, when deserializing an argument we know the type. Therefore wrapping the json object with the type would not be needed. Enums could just be integers.
And the `SerializationManager` would not be needed any more.
So the generated code could do something like `Class.deserialize(jsonDecodedData)`. The same applies to deserializing method return types at the client side and the whole future call thematic.

However I'm sure unexpected challenges will appear while implementing it.

**Reply by vlidholt** [reply]
In the protocol with the database and between client and server, the types are known. But they may not be known in caching or future calls. One possible solution could be to just wrap the top-level object with the class-name.

JSON vs Protobuf seem to have different advantages/disadvantages. JSON is easier to read and understand, Protobuf is more efficient. JSON may also be needed to store things in the database. For future versions, I'm also planning on supporting typed database documents (similar to Firestore, but based on a structure provided by the yaml-files). That would probably required a JSON solution.

So an optimal solution would possibly be to have both options.

**Reply by fischerscode** [reply]
It's slowly comming together. I hope I'll finish a first PR attempt tomorrow.

**Reply by vlidholt** [reply]
Hi! I saw the PR. Unfortunately, I won't have time to look through it in detail until Friday this week as I'm preparing a large Flutter/Serverpod event in Stockholm on Thursday. Overall, your work looks very cool. Love the idea of making the serialization compatible with Freezed and many of your other improvements.

Would you be available for a quick chat on video sometime next days? Would love to synch up. Feel free to add me on Linkedin and we can take it from there: https://www.linkedin.com/in/viktorlidholt/

**Reply by vlidholt** [reply]
@fischerscode 👆
